haystack/components/generators/chat/hugging_face_local.py

```diff
@@ -0,0 +1,366 @@
+import logging
+import sys
+from typing import Any, Dict, List, Literal, Optional, Union, Callable
+
+from haystack.components.generators.hf_utils import PIPELINE_SUPPORTED_TASKS
+
+from haystack import component, default_to_dict, default_from_dict
+from haystack.components.generators.hf_utils import HFTokenStreamingHandler
+from haystack.components.generators.utils import serialize_callback_handler, deserialize_callback_handler
+from haystack.dataclasses import ChatMessage, StreamingChunk
+from haystack.lazy_imports import LazyImport
+
+logger = logging.getLogger(__name__)
+
+with LazyImport(message="Run 'pip install transformers[torch]'") as torch_and_transformers_import:
+    import torch
+    from huggingface_hub import model_info
+    from transformers import StoppingCriteriaList, pipeline, PreTrainedTokenizer, PreTrainedTokenizerFast
+    from haystack.components.generators.hf_utils import StopWordsCriteria  # pylint: disable=ungrouped-imports
+
+
+@component
+class HuggingFaceLocalChatGenerator:
+    """
+
+    The `HuggingFaceLocalChatGenerator` class is a component designed for generating chat responses using models from
+    Hugging Face's model hub. It is tailored for local runtime text generation tasks and provides a convenient interface
+    for working with chat-based models, such as `mistralai/Mistral-7B-Instruct-v0.2` or `meta-llama/Llama-2-7b-chat-hf`
+    etc.
+
+    Usage example:
+    ```python
+    from haystack.components.generators.chat import HuggingFaceLocalChatGenerator
+    from haystack.dataclasses import ChatMessage
+
+
+
+    generator = HuggingFaceLocalChatGenerator(model="mistralai/Mistral-7B-Instruct-v0.2", device="cuda:0")
+    messages = [ChatMessage.from_user("What's Natural Language Processing? Be brief.")]
+    print(generator.run(messages))
+
+    # {'replies': [ChatMessage(content=' Natural Language Processing (NLP) is a subfield of artificial
+    intelligence that deals with the interaction between computers and human language. It enables computers
+    to understand, interpret, and generate human language in a valuable way. NLP involves various techniques
+    such as speech recognition, text analysis, sentiment analysis, and machine translation. The ultimate goal
+    is to make it easier for computers to process and derive meaning from human language, improving communication
+    between humans and machines.', role=<ChatRole.ASSISTANT: 'assistant'>, name=None,
+    meta={'finish_reason': 'stop', 'index': 0, 'model': 'mistralai/Mistral-7B-Instruct-v0.2',
+    'usage': {'completion_tokens': 90, 'prompt_tokens': 19, 'total_tokens': 109}})]}
+    ```
+    """
+
+    def __init__(
+        self,
+        model: str = "mistralai/Mistral-7B-Instruct-v0.2",
+        task: Optional[Literal["text-generation", "text2text-generation"]] = None,
+        device: Optional[str] = None,
```

Comment:
the device management should be refactored, as done in the HFLocalGenerator: https://github.com/deepset-ai/haystack/blob/a7ac4edd074f75171145342803d42e8639c2d23c/haystack/components/generators/hugging_face_local.py#L46